### PR TITLE
Adds ScopeDecorator and backports log correlation to use it

### PR DESCRIPTION
**Pull request by adriancole**
This is based on some work in Fukuoka where @anuraaga noticed we can stop some of the subclassing we do now

### Customizing in-process propagation (Ex. log correlation)

The `CurrentTraceContext` makes a trace context visible, such that spans
aren't accidentally added to the wrong trace. It also accepts hooks like
log correlation.

For example, if you use Log4J 2, you can copy trace IDs to your logging
context with [our decorator](https://github.com/openzipkin/brave/tree/master/context/log4j2):
```java
tracing = Tracing.newBuilder()
    .currentTraceContext(ThreadLocalCurrentTraceContext.newBuilder()
       .addScopeDecorator(ThreadContextScopeDecorator.create())
       .build()
    )
    ...
    .build();
```

Besides logging, other tools are available. `StrictScopeDecorator` can
help find out when code is not closing scopes properly. This can be
useful when writing or diagnosing custom instrumentation.

See #682

### CODE REVIEW

File: brave/src/main/java/brave/internal/propagation/CorrelationFieldScopeDecorator.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    } else {
+      boolean sameParentId = lowerHexEqualsUnsignedLong(previousParentId, parentId);
+      if (!sameParentId) put("parentId", HexCodec.toLowerHex(parentId));
+    }
+
+    boolean sameSpanId = lowerHexEqualsUnsignedLong(previousSpanId, currentSpan.spanId());
+    if (!sameSpanId) put("spanId", HexCodec.toLowerHex(currentSpan.spanId()));
+  }
+
+  protected abstract @Nullable String getIfString(String key);
```
Comment (anuraaga) [inline brave/src/main/java/brave/internal/propagation/CorrelationFieldScopeDecorator.java]: Javadoc? Also I'm not sure how to grok this `get` - at first I expected it to get the value from the current trace context. `getPreviouslyCorrelated` (a bit wordy though)?

File: brave/src/main/java/brave/propagation/CurrentTraceContext.java [hunk 2]
```diff
@@ -21,23 +24,82 @@
...
+  public interface Factory {
+    CurrentTraceContext create(List<ScopeDecorator> scopeDecorators);
+  }
+
+  protected CurrentTraceContext() {
+    this.scopeDecorators = Collections.emptyList();
+  }
+
+  protected CurrentTraceContext(Builder<?> builder) {
+    this.scopeDecorators = (List<ScopeDecorator>) builder.scopeDecorators.clone();
```
Comment (anuraaga) [inline brave/src/main/java/brave/propagation/CurrentTraceContext.java]: I personally haven't seen `clone()` for copying anything other than an array. `new ArrayList<ScopeDecorator>(builder.scopeDecorators)`?

File: brave/src/main/java/brave/propagation/CurrentTraceContext.java [hunk 3]
```diff
@@ -21,23 +24,82 @@
...
+   *     }
+   *     Scope result = new ThreadLocalScope();
+   *     // ensure scope hooks are attached to the result
+   *     return decorateScope(currentSpan, result);
+   *   }
+   * }</pre>
+   */
+  protected Scope decorateScope(@Nullable TraceContext currentSpan, Scope scope) {
+    int length = scopeDecorators.size();
+    if (length == 0) return scope;
```
Comment (anuraaga) [inline brave/src/main/java/brave/propagation/CurrentTraceContext.java]: I think the for loop duplicates this check so it doesn't really speed up the `0` case.

**Comment by adriancole** [comment]
will be used by https://github.com/line/armeria/pull/1328

File: brave/src/main/java/brave/internal/propagation/CorrelationFieldScopeDecorator.java (hunk 1 above)
Comment (adriancole) [reply brave/src/main/java/brave/internal/propagation/CorrelationFieldScopeDecorator.java]: yep. it is internal, but no excuse :P

File: brave/src/main/java/brave/propagation/CurrentTraceContext.java (hunk 3 above)
Comment (adriancole) [reply brave/src/main/java/brave/propagation/CurrentTraceContext.java]: agreed

File: brave/src/main/java/brave/internal/propagation/CorrelationFieldScopeDecorator.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    boolean sameSpanId = lowerHexEqualsUnsignedLong(previousSpanId, currentSpan.spanId());
+    if (!sameSpanId) put("spanId", HexCodec.toLowerHex(currentSpan.spanId()));
+  }
+
+  /**
+   * Returns the correlation property of the specified name iff it is a string, or null otherwise.
+   */
+  protected abstract @Nullable String getIfString(String key);
```
Comment (anuraaga) [inline brave/src/main/java/brave/internal/propagation/CorrelationFieldScopeDecorator.java]: `ifString` seems like it was an implementation detail of `MDC`, for this abstract class, how about just `get`?

Comment (adriancole) [reply brave/src/main/java/brave/internal/propagation/CorrelationFieldScopeDecorator.java]: ok with me!
